components-core/src/main/java/com/adyen/checkout/components/repository/PublicKeyRepository.kt

```diff
@@ -10,28 +10,40 @@ package com.adyen.checkout.components.repository
 
 import com.adyen.checkout.components.api.PublicKeyService
 import com.adyen.checkout.core.api.Environment
-import com.adyen.checkout.core.exception.CheckoutException
 import com.adyen.checkout.core.log.LogUtil
 import com.adyen.checkout.core.log.Logger
-import org.json.JSONException
-import java.io.IOException
-
-private val TAG = LogUtil.getTag()
-private const val CONNECTION_RETRIES = 3
+import com.adyen.checkout.core.util.runSuspendCatching
 
 class PublicKeyRepository {
 
-    suspend fun fetchPublicKey(environment: Environment, clientKey: String): String {
+    suspend fun fetchPublicKey(
+        environment: Environment,
+        clientKey: String
+    ): Result<String> = runSuspendCatching {
         Logger.d(TAG, "fetching publicKey from API")
-        repeat(CONNECTION_RETRIES) {
+
+        retryOnFailure(CONNECTION_RETRIES) {
+            PublicKeyService(environment).getPublicKey(clientKey).publicKey
+        }
+    }
+
+    @Suppress("SameParameterValue")
+    private inline fun <T> retryOnFailure(times: Int, block: () -> T): T {
+        repeat(times - 1) {
```

Comment:
Why should this be `times - 1` instead of `times`?

Reply:
Never mind, just saw the extra `return block()`, I'm not sure if this is obvious and I missed it or we should improve it somehow

Reply:
Let me explain with an example. You do `retryOnFailure(3) {}`. The repeat statement will try the block `3-1=2` times ignoring any exceptions. If it fails both times, then the last line of the method `return block()` will try one last time, throwing an exception if it fails. So in total the block has been executed 3 times

Reply:
Does this look better or worse? 😅 
```kotlin
    @Suppress("SameParameterValue")
    private inline fun <T> retryOnFailure(times: Int, block: () -> T): T {
        var throwable = Throwable()
        repeat(times) { index ->
            @Suppress("TooGenericExceptionCaught")
            try {
                return block()
            } catch (e: Throwable) {
                throwable = e
            }
        }
        throw  throwable
    }
```

Reply:
Your logic is sound and definitely works, I'm only wondering if `repeat(times - 1)` is confusing because it made me think `repeat` is zero based or something. Still I can't write it in a "cleaner" way because the compiler will complain that we aren't returning a value in all paths.

Reply:
It's easier to understand, so I like it 👍 

Small adjustment: `var throwable = Throwable()` -> `lateinit var throwable: Throwable`

Reply:
Haha amazing, that's actually one of the few times I see `lateinit` used correctly outside of the usual init during lifecycle use case 👏